### PR TITLE
add deb build and modify travis config to support release works

### PR DESCRIPTION
**Pull request by changev**
### add deb build

HWIMO-BUILD can build on-static-common.deb, which contains common and pxe static files built from build_all.sh.
#### used

per PR-merge build static files and upload them to bintray and overwrite the old ones.
#### now
1. per PR-merge build on-static-common_1.2.3~rc-20160923181601Z-125b07b.deb or on-static-common_1.2.2~devel-20150923181601Z-535b07b.deb with deb_component "main", we call this ci-builds.
2. per push Tag build  on-static-common_1.2.3.deb with deb_component "release", we call this release build

This feature will be applied on every repo in the future.
For example, in on-tftp : https://github.com/RackHD/on-tftp/pull/53
@RackHD/corecommitters  @panpan0000 @PengTian0

### CODE REVIEW

File: .bintray.json.in [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    },
+
+    "version": {
+        "name": "#REVISION#"
+    },
+
+    "files":
+        [
+            {
+                "includePattern": "deb/(.*)", "uploadPattern": "/dists/trusty/main/binary-amd64/$1",
```
Comment (panpan0000) [inline .bintray.json.in]: if we specific trusty/main/amd64, why we still need to specific the folder /dists/trusty/main/binary-amd64 ?
say, can we just "uploadPattern": "$1" ?

File: .travis.yml [hunk 2]
```diff
@@ -11,13 +11,15 @@ before_deploy:
...
+- ./extra/make-cfg.sh
+- ./extra/make-deb.sh
 deploy:
   provider: bintray
   file: .bintray.json
   user: $BINTRAY_USER
   key: $BINTRAY_KEY
   on:
-    all_branches: true
+    condition: $TRAVIS_TAG =~ ^[0-9.]+$ || -n $TRAVIS_BRANCH
```
Comment (panpan0000) [inline .travis.yml]: Does it mean only trigger on when new tag created ? how about master daily build ? will it trigger Travis CI?

File: HWIMO-BUILD [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+# HWIMO-BUILD
+
+#
+# Build debian package for all images created by on-imagebuilder
+#
+set -e
+output_path=/tmp/on-imagebuilder
```
Comment (panpan0000) [inline HWIMO-BUILD:9]: DO you need mkdir -p first ?

File: HWIMO-BUILD [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+if [ -n "$BUILD_NUMBER" ]
+then
+  PKG_VERSION="${PKG_VERSION}-${BUILD_NUMBER}"
+fi
+
+export DEBEMAIL="hwimo robots <hwimo@hwimo.lab.emc.com>"
+export DEBFULLNAME="The HWIMO Robots"
+
+# build files in packagebuild into a debian package 
+dch -v ${PKG_VERSION} autobuild
```
Comment (panpan0000) [inline HWIMO-BUILD:45]: "-v" will ignore the version in changelog. it means "--newversion"

File: debian/changelog.dch [hunk 5]
```diff
@@ -0,0 +1,9 @@
+on-static-common (1.1.1) unstable; urgency=low
```
Comment (panpan0000) [inline debian/changelog.dch]: what does dch file mean? is itgenerated in runtime ?

File: debian/control [hunk 6]
```diff
@@ -0,0 +1,13 @@
+Source: on-static-common
```
Comment (panpan0000) [inline debian/control]: I just have an idea :   should we call it on-imagebuilder , to align the repo name ?
Or you think on-static is better ?

File: debian/on-static-common.install [hunk 7]
```diff
@@ -0,0 +1,2 @@
+common var/renasar/on-http/static/http
```
Comment (panpan0000) [inline debian/on-static-common.install]: Nice

File: extra/make-cfg.sh [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
+#  This script will generate config for make-deb and deployment on bintray in TravisCI, 
+#  This script can't work without TravisCI
```
Comment (panpan0000) [inline extra/make-cfg.sh]: please specific an reason. for example, requires some ENV Variable ?

File: extra/make-cfg.sh [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+# Ensure we're always in the right directory.
+SCRIPT_DIR="$( cd "$( dirname "$0" )" && pwd )"
+cd $SCRIPT_DIR/..
+
+#Generate SYMBOL according to branch or tag
+if [ -n "$TRAVIS_BRANCH" ]; then
+   RRANCH=${TRAVIS_BRANCH}
```
Comment (panpan0000) [inline extra/make-cfg.sh]: TYPO BRANCH

File: extra/make-cfg.sh [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+echo "DEBDIR:       $DEBDIR"
+echo "DEBFULLNAME:  $DEBFULLNAME"
+echo "DEBEMAIL:     $DEBEMAIL"
+echo "DEBBRANCH:    $DEBBRANCH"
+echo "DEBPKGVER:    $DEBPKGVER"
+echo "DCHOPTS:      $DCHOPTS"
+
+#modify bintray cfg
+VERSION=`dpkg-parsechangelog --show-field Version`
+REVISION=${VERSION}${DEBBRANCH}
```
Comment (panpan0000) [inline extra/make-cfg.sh]: what's the deliminator  between ?  debian only recognize "~"
and this variable DEBBRANCH doesn't contain only branch info , but actually, commit info ,can we rename it more meaningful ?

File: extra/make-cfg.sh [hunk 11]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
+#  This script will generate config for make-deb and deployment on bintray in TravisCI,
```
Comment (panpan0000) [inline extra/make-cfg.sh]: what kind of config ?
actually,  "make-deb" only use the exported DEBBRANCH variable, the config(.bintray.json) is used for TravisCI

File: extra/make-cfg.sh [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+if [ -z "$DEBBRANCH" ] && [ "$SYMBOL" != "release" ]; then
+        export DEBBRANCH=`echo "${SYMBOL}-${DATESTRING}" | sed 's/[\/\_]/-/g'`
+fi
+
+if [ -z "$DEBPKGVER" ]; then
+  export DEBPKGVER=`git log -n 1 --pretty=oneline --abbrev-commit`
+fi
+
+if [ -z "$DCHOPTS" ]; then
+        export DCHOPTS="-l ${DEBBRANCH} -u low ${DEBPKGVER}"
```
Comment (panpan0000) [inline extra/make-cfg.sh]: actually, I dislike those export ...
unnecessary variable pollution  ...

**Comment by panpan0000** [comment]
moreover, TravisCI fails ?

File: .travis.yml (hunk 2 above)
Comment (changev) [reply .travis.yml]: "merge event" and "push tag event" will all trigger build and deploy.

File: HWIMO-BUILD (hunk 3 above)
Comment (changev) [reply HWIMO-BUILD:9]: Don't need that, build_all will do that.

File: HWIMO-BUILD (hunk 4 above)
Comment (changev) [reply HWIMO-BUILD:45]: Yes, format and function of this `HWIMO-BUILD` here is totally the same with that in other repos. IMO HWIMO-BUILD is just to build a deb ignore version information. 
Should anyone run HWIMO-BUILD and build a deb with correct version?
If so I'll add this change to HWIMO-BUILD in all repos.

File: debian/changelog.dch (hunk 5 above)
Comment (changev) [reply debian/changelog.dch]: Ignore this please, this file was unexpected created when I tried dch cmds, I'll delete this.

File: debian/control (hunk 6 above)
Comment (changev) [reply debian/control]: IMO "builder" is inappropriate for a deb. on-static-common is inherited from Onrack and "common" means there're specific static files too. So I think it's a proper name.

File: extra/make-cfg.sh (hunk 11 above)
Comment (changev) [reply extra/make-cfg.sh]: maybe make-env.sh is more proper? What's your advice?

File: extra/make-cfg.sh (hunk 8 above)
Comment (changev) [reply extra/make-cfg.sh]: Thanks, I'll add more explaination

File: extra/make-cfg.sh (hunk 12 above)
Comment (changev) [reply extra/make-cfg.sh]: In make-deb.sh it also "export variables", so I think "env vars" is a concise way to pass vars to make-deb.sh.

and make-deb.sh can run stand-alone without any other scripts.

**Comment by changev** [comment]
@panpan0000 fatal: unable to access 'https://git.kernel.org/pub/scm/boot/syslinux/syslinux.git/': gnutls_handshake() failed: Handshake failed

File: extra/make-deb.sh [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        export DEBBRANCH=$(./extra/gen-debbranch.sh)
+fi
+
+if [ -z "$DEBPKGVER" ]; then
+        export DEBPKGVER=`git log -n 1 --pretty=oneline --abbrev-commit`
+fi
+
+if [ -z "$DCHOPTS" ]; then
+        if [[ $DEBBRANCH =~ ^[0-9.]+$ ]]; then
+                export DCHOPTS="-r \"\""
```
Comment (panpan0000) [inline extra/make-deb.sh:32]: comments is better , to explain those special parameter like -r .

File: debian/control (hunk 6 above)
Comment (rackhd-test-rel) [reply debian/control]: ok

**Comment by rackhd-test-rel** [comment]
fine to me , now :+1:

File: .bintray-deb.json.in [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    },
+
+    "version": {
+        "name": "#REVISION#"
+    },
+
+    "files":
+        [
+            {
+                "includePattern": "deb/(.*)", "uploadPattern": "$1",
```
Comment (yyscamper) [inline .bintray-deb.json.in]: A style suggestion: move `"uploadPattern": "$1"` to a new line.

File: .bintray-gen.json [hunk 15]
```diff
@@ -25,4 +25,4 @@
             }
```
Comment (yyscamper) [inline .bintray-gen.json:25]: This file change is not necessary, and the file mode change from 644 to 755 is also unnecessary.

**Review by yyscamper — APPROVED** [review]

**Pull request merged**